lgc/patch/PatchReadfirstlaneOpt.h

```diff
@@ -0,0 +1,63 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  PatchReadfirstlaneOpt.h
+ * @brief LLPC header file: contains declaration of class lgc::PatchReadfirstlaneOpt.
+ ***********************************************************************************************************************
+ */
+#pragma once
+
+#include "lgc/util/Internal.h"
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/Pass.h"
+
+namespace llvm {
+class LegacyDivergenceAnalysis;
+} // namespace llvm
+
+namespace lgc {
+class PatchReadfirstlaneOpt final : public llvm::FunctionPass, public llvm::InstVisitor<PatchReadfirstlaneOpt> {
+public:
+  PatchReadfirstlaneOpt();
+
+  bool runOnFunction(llvm::Function &function) override;
+  void getAnalysisUsage(llvm::AnalysisUsage &analysisUsage) const override;
+
+  void visitCallInst(llvm::CallInst &callInst);
+
+  static char ID; // ID of this pass
+
+private:
+  PatchReadfirstlaneOpt(const PatchReadfirstlaneOpt &) = delete;
+  PatchReadfirstlaneOpt &operator=(const PatchReadfirstlaneOpt &) = delete;
+
+  bool pushUpReadfirstlane();
+
+  llvm::LegacyDivergenceAnalysis *m_divergenceAnalysis; // The divergence analysis.
+  llvm::DenseSet<llvm::Value *> m_intialUniforms;       //
+};
```

Comment:
Please move the definition of this class into the .cpp file. LLVM style is to usually have pass definitions in the .cpp file if the implementation fits in a single file. See other passes like `PatchEntryPointMutate` for how this is done.

Reply:
Thanks for telling the rule. Done.